File: src/main/java/editor/database/card/ModalCard.java
package editor.database.card;

/**
 * This class represents a Card with two faces: One on the front, and one on the
 * back. During a game, one side can be cast.
 *
 * @author Alec Roelke
 */
public class ModalCard extends MultiCard
{
    /**
     * Card representing the front face.
     */
    private final Card front;

    /**
     * Create a new ModalCard with the given Cards as faces.  Their layouts should
     * say that they are modal double-faced cards.
     *
     * @param f card representing the front face
     * @param b card representing the back face
     */
    public ModalCard(Card f, Card b)
    {
        super(CardLayout.MODAL_DFC, f, b);
        front = f;
        if (front.layout() != CardLayout.MODAL_DFC || b.layout() != CardLayout.MODAL_DFC)
            throw new IllegalArgumentException("can't join non-modal-double-faced cards into modal double-faced cards");
    }

    /**
     * {@inheritDoc}
     * The mana value of a modal card is that of the front face (until you cast it or it enters the battlefield,
     * at which point it becomes that of the side facing up).
     */
    public double manaValue()
    {
        return front.manaValue();
    }
}

File: src/main/scala/editor/database/card/ModalCard.scala
package editor.database.card

import CardLayout._

/**
 * A [[Card]] with two faces, either of which can be played, and doesn't typically switch until it
 * leaves the battlefield or stack. Its mana value is that of its front face.
 * 
 * @constructor create a new modal double-faced card
 * @param front front face of the card
 * @param back back face of the card
 * @author Alec Roelke
 */
@throws[IllegalArgumentException]("if both faces aren't modal double-faced cards")
class ModalCard(front: Card, back: Card) extends MultiCard(MODAL_DFC, IndexedSeq(front, back)) {
  if (front.layout != MODAL_DFC || back.layout != MODAL_DFC)
    IllegalArgumentException("can't join non-modal-double-faced cards into modal double-faced cards")

  override def manaValue = front.manaValue
  override def minManaValue = faces.map(_.manaValue).min
  override def maxManaValue = faces.map(_.manaValue).max
  override def avgManaValue = faces.map(_.manaValue).sum/faces.size
  override lazy val imageNames = front.imageNames ++ back.imageNames
}